rootston/keyboard.c

```diff
@@ -10,32 +11,95 @@
 #include <xkbcommon/xkbcommon.h>
 #include "rootston/input.h"
 
+static ssize_t keyboard_pressed_keysym_index(struct roots_keyboard *keyboard,
+		xkb_keysym_t keysym) {
+	for (size_t i = 0; i < ROOTS_KEYBOARD_PRESSED_KEYSYMS_CAP; i++) {
+		if (keyboard->pressed_keysyms[i] == keysym) {
+			return i;
+		}
+	}
+	return -1;
+}
+
+static void keyboard_binding_execute(struct roots_keyboard *keyboard,
+		char *command) {
+	struct roots_server *server = keyboard->input->server;
+	if (strcmp(command, "exit") == 0) {
+		wl_display_terminate(server->wl_display);
+	} else {
+		pid_t pid = fork();
+		if (pid < 0) {
+			wlr_log(L_ERROR, "cannot execute binding command: fork() failed");
+			return;
+		} else if (pid == 0) {
+			execl("/bin/sh", "/bin/sh", "-c", command, (void *)NULL);
```

Comment:
You'll probably want to close FDs here, right

Reply:
Ah, right

Reply:
It's also reasonable to just double check CLOEXEC. Up to you.